cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go

```diff
@@ -96,31 +96,41 @@ func addNodeSelectorToPodSpec(podSpec *v1.PodSpec) {
 // setMasterTolerationOnPodSpec makes the Pod tolerate the master taint
 func setMasterTolerationOnPodSpec(podSpec *v1.PodSpec) {
 	if podSpec.Tolerations == nil {
-		podSpec.Tolerations = []v1.Toleration{kubeadmconstants.MasterToleration}
+		podSpec.Tolerations = []v1.Toleration{
+			kubeadmconstants.MasterToleration,
+			kubeadmconstants.CloudProviderUninitializedToleration,
+		}
 		return
 	}
 
-	podSpec.Tolerations = append(podSpec.Tolerations, kubeadmconstants.MasterToleration)
+	podSpec.Tolerations = append(podSpec.Tolerations,
+		kubeadmconstants.MasterToleration,
+		kubeadmconstants.CloudProviderUninitializedToleration)
 }
 
 // setHostIPOnPodSpec sets the environment variable HOST_IP using downward API
 func setHostIPOnPodSpec(podSpec *v1.PodSpec) {
-	envVar := v1.EnvVar{
-		Name: "HOST_IP",
-		ValueFrom: &v1.EnvVarSource{
-			FieldRef: &v1.ObjectFieldSelector{
-				FieldPath: "status.hostIP",
-			},
-		},
-	}
-
-	podSpec.Containers[0].Env = append(podSpec.Containers[0].Env, envVar)
-
 	for i := range podSpec.Containers[0].Command {
 		if strings.Contains(podSpec.Containers[0].Command[i], "advertise-address") {
-			podSpec.Containers[0].Command[i] = "--advertise-address=$(HOST_IP)"
+			podSpec.Containers[0].Command[i] = ""
```

Comment:
Why changing this and the code above and below in the function?

Reply:
HOST_IP  need node.spec.status.address to function well. But under external mode, untill cloudprovider is initialized, there is no address exists in pod status. this would result in daemonset creation failure with error of ```host IP unknown; known addresses: ```.

Reply:
See kubelet_node_status.go  setNodeAddress
```
// Set IP and hostname addresses for the node.
func (kl *Kubelet) setNodeAddress(node *v1.Node) error {
	if kl.nodeIP != nil {
		if err := validateNodeIP(kl.nodeIP); err != nil {
			return fmt.Errorf("failed to validate nodeIP: %v", err)
		}
		glog.V(2).Infof("Using node IP: %q", kl.nodeIP.String())
	}

	if kl.externalCloudProvider {
		if kl.nodeIP != nil {
			if node.ObjectMeta.Annotations == nil {
				node.ObjectMeta.Annotations = make(map[string]string)
			}
			node.ObjectMeta.Annotations[kubeletapis.AnnotationProvidedIPAddr] = kl.nodeIP.String()
		}
		// We rely on the external cloud provider to supply the addresses.
		return nil
	}
```

Reply:
/cc @andrewsykim @fabriziopandini @luxas @timothysc

Reply:
@andrewsykim @fabriziopandini @luxas @timothysc   Any comments? Thanks

Reply:
The node address status is indeed a limitation of using external cloud providers that has not been addressed yet (as far as I know). A couple things to note though:
*  `--cloud-provider=external` doesn't need to be set for kubelets that run the control plane, hence you would use whatever the default node address is. 
*  `--cloud-provider=external` isn't even a valid cloud provider on kubeadm yet (see https://github.com/kubernetes/kubernetes/pull/50545), so should we be accounting for it anyways?

Reply:
I don't have any strong opinions on either direction, but ^ those were considerations in the original PR :).

Reply:
```--cloud-provider=exernal ``` has been accepted as a valid cloud provider.  https://github.com/kubernetes/kubernetes/commit/f302307437c4f1b38d43e72567f2630f3e53431f
Or, we could remove ```advertise-address``` temporarily until we have a better solution for it? Hence control planes better to have consistent configuration about external cloudprovider.

Reply:
seems like this can be fixed by using POD_IP

Reply:
kubelet_pods.go generateAPIPodStatus()
```
	if kl.kubeClient != nil {
		hostIP, err := kl.getHostIPAnyWay()
		if err != nil {
			glog.V(4).Infof("Cannot get host IP: %v", err)
		} else {
			s.HostIP = hostIP.String()
			if kubecontainer.IsHostNetworkPod(pod) && s.PodIP == "" {
				s.PodIP = hostIP.String()
			}
		}
	}
```

Under ```hostNetwork: true``` mode, podIP comes from kl.getHostIPAnyWay() which also updated by cloudprovider.

Reply:
@andrewsykim  @xiangpengzhao  What if we just remove the ```--advertise-address=``` configuration?